ssl.go

```diff
@@ -95,16 +99,72 @@ func sslLoadCert(ctx *openssl.Ctx, certFile string) (err error) {
 	return
 }
 
-func sslLoadKey(ctx *openssl.Ctx, keyFile string) (err error) {
+func sslLoadKey(ctx *openssl.Ctx, keyFile string, password string,
+	passwordFile string) error {
 	var keyBytes []byte
+	var err error
+
 	if keyBytes, err = ioutil.ReadFile(keyFile); err != nil {
-		return
+		return err
 	}
 
 	var key openssl.PrivateKey
-	if key, err = openssl.LoadPrivateKeyFromPEM(keyBytes); err != nil {
-		return
+	var errs []error
+
+	// If the key is encrypted and password is not provided,
+	// openssl.LoadPrivateKeyFromPEM(keyBytes) asks to enter PEM pass phrase
+	// interactively. On the other hand,
+	// openssl.LoadPrivateKeyFromPEMWithPassword(keyBytes, password) works fine
+	// for non-encrypted key with any password, including empty string. If
+	// the key is encrypted, we fast fail with password error instead of
+	// requesting the pass phrase interactively.
+	key, err = openssl.LoadPrivateKeyFromPEMWithPassword(keyBytes, password)
+	if err == nil {
+		return ctx.UsePrivateKey(key)
+	} else {
+		errs = append(errs, err)
+	}
+
+	if passwordFile != "" {
+		var file *os.File
+		file, err = os.Open(passwordFile)
+		if err == nil {
+			defer file.Close()
+
+			scanner := bufio.NewScanner(file)
+			// Tarantool itself tries each password file line.
+			for scanner.Scan() {
+				password = strings.TrimSpace(scanner.Text())
+
+				key, err = openssl.LoadPrivateKeyFromPEMWithPassword(keyBytes, password)
+				if err == nil {
+					return ctx.UsePrivateKey(key)
+				} else {
+					errs = append(errs, err)
+				}
+			}
+		} else {
+			errs = append(errs, err)
+		}
+	}
+
+	if len(errs) > 1 {
+		// Convenient multiple error wrapping was introduced only in Go 1.20
+		// https://pkg.go.dev/errors#example-Join
+		// https://github.com/golang/go/issues/53435
+		rerr := errors.New("got multiple errors on SSL decryption")
+		var i int
+		for i, err = range errs {
+			if i == 0 {
+				// gofmt forbids error strings to end with punctuation or newlines
+				rerr = fmt.Errorf("%s: %w", rerr, err)
```

Comment:
Could you please show an example of the resulting error? It seems to me that the sequence of errors will not give much more information.

Reply:
For `inv_pass_and_inv_passfile_key_encrypt_client`:

```
got multiple errors on SSL decryption: failed reading private key: SSL errors: Provider routines::bad decrypt
PKCS12 routines::pkcs12 cipherfinal error
Provider routines::bad decrypt
PKCS12 routines::pkcs12 cipherfinal error, failed reading private key: SSL errors: Provider routines::bad decrypt
PKCS12 routines::pkcs12 cipherfinal error
Provider routines::bad decrypt
PKCS12 routines::pkcs12 cipherfinal error
```

For invalid password and non-existing file
```
got multiple errors on SSL decryption: failed reading private key: SSL errors: Provider routines::bad decrypt
PKCS12 routines::pkcs12 cipherfinal error
Provider routines::bad decrypt
PKCS12 routines::pkcs12 cipherfinal error, open testdata/invalidpasswords1: no such file or directory
```

Reply:
Maybe just return the last error? There will be either an error reading the file, or the fact that the password is not match.

Reply:
Updated. (But now it's first error instead of the last one since we read file before decrypting. Tarantool behaves the same: if password is good and file not exist, it is a success decryption, if decryption fails and file not exists -- we return a file error.)

Reply:
Thank you!